// react-pubsub/lib/adapter/createSubscription.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
/*
   Creates the subscription object

   @param autoUnmount boolean defaults to true, if it has to remove all subscriptions on cwunmount
   @param adapter object the pubsub adapter
   @return the subscription object
   */
var createSubscription = function createSubscription(adapter, unsubscribe) {
  if (!adapter) {
    throw new Error('\'createSubscription()\' expected an adapter as first argument instead received: ' + adapter);
  }

  if (!unsubscribe) {
    throw new Error('\'createSubscription()\' expected a function as second argument instead ' + ('received: ' + unsubscribe));
  }

  var sub = {
    unsubscribe: unsubscribe,
    subscriptions: [],

    add: function add(action, cb) {
      if (!action || typeof action !== 'string') {
        throw new Error('Subscription \'.add()\' expected an action (string) as first argument ' + ('instead received: ' + action));
      }

      if (!cb || typeof cb !== 'function') {
        throw new Error('Subscription \'.add()\' expected a function as second argument instead received: ' + cb);
      }

      var token = adapter.subscribe(action, cb);
      sub.subscriptions.push(token);
      return function () {
        var idx = sub.subscriptions.indexOf(token);
        if (idx < 0) {
          throw new Error('PubSub Error: you\'re trying to unsubscribe an unrecognized token');
        }
        sub.subscriptions.splice(idx, 1);
        adapter.unsubscribe(token);
      };
    },
    removeAll: function removeAll() {
      sub.subscriptions.forEach(function (token) {
        return adapter.unsubscribe(token);
      });
      sub.subscriptions = [];
    },
    publish: function publish(action) {
      if (!action || typeof action !== 'string') {
        throw new Error('Subscription \'.publish()\' expected an action (string) as first argument ' + ('instead received: ' + action));
      }

      for (var _len = arguments.length, params = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        params[_key - 1] = arguments[_key];
      }

      adapter.publish.apply(adapter, [action].concat(params));
    }
  };

  return sub;
};

exports.default = createSubscription;
