// react-pubsub/lib/adapter/createDefaultAdapter.js
'use strict';

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = void 0; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

Object.defineProperty(exports, "__esModule", {
  value: true
});
var createDefaultAdapter = function createDefaultAdapter() {
  var actions = {};
  var hOP = actions.hasOwnProperty;
  return {
    subscribe: function subscribe(action, cb) {
      if (!action || typeof action !== 'string') {
        throw new Error('Default adapter \'.subscribe()\' ' + 'expected an action (string) as first argument instead ' + ('received: ' + action));
      }

      if (!cb || typeof cb !== 'function') {
        throw new Error('Default adapter \'.subscribe()\' expected a function as second ' + ('argument instead received: ' + cb));
      }

      if (!hOP.call(actions, action)) {
        actions[action] = [];
      }

      actions[action].push(cb);
      return [action, cb];
    },
    unsubscribe: function unsubscribe(token) {
      if (!token || !Array.isArray(token)) {
        throw new Error('Default adapter \'.unsubscribe()\' expected a ' + ('token (array) instead received: ' + token));
      }

      if (token.length !== 2 || typeof token[0] !== 'string' || typeof token[1] !== 'function') {
        throw new Error('Default adapter \'.unsubscribe()\' expected a valid token ' + ('([string, function]) instead received: ' + token));
      }

      var _token = _slicedToArray(token, 2);

      var action = _token[0];
      var cb = _token[1];

      var callbacks = actions[action] || [];
      var idx = callbacks.indexOf(cb);
      if (idx < 0) {
        console.error( // eslint-disable-line no-console
        'You\'re unsubscribing an unrecognized token ' + token);
      } else {
        callbacks.splice(idx, 1);
      }
    },
    publish: function publish(action) {
      for (var _len = arguments.length, params = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        params[_key - 1] = arguments[_key];
      }

      if (!action || typeof action !== 'string') {
        throw new Error('Default adapter \'.publish()\' expected an action (string) as ' + ('first argument instead received: ' + action));
      }

      if (!hOP.call(actions, action)) {
        console.info( // eslint-disable-line no-console
        'The action \'' + action + '\' being published has no listeners');
        return;
      }

      actions[action].forEach(function (cb) {
        return cb.apply(void 0, params);
      });
    }
  };
};

exports.default = createDefaultAdapter;
