// react-pubsub/lib/components/PubSubProvider.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _pubSubShape = require('../shapes/pubSubShape');

var _pubSubShape2 = _interopRequireDefault(_pubSubShape);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var didWarnAboutCoreChanged = false;
var warnOfCoreChanges = function warnOfCoreChanges() {
  if (didWarnAboutCoreChanged) {
    return;
  }
  didWarnAboutCoreChanged = true;
  console.error( // eslint-disable-line no-console
  '<PubSubProvider> currently can\'t dynamically change the pubSubCore!');
};

var PubSubProvider = function (_Component) {
  _inherits(PubSubProvider, _Component);

  function PubSubProvider(props, context) {
    _classCallCheck(this, PubSubProvider);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(PubSubProvider).call(this, props, context));

    _this.pubSubCore = props.pubSubCore;
    return _this;
  }

  _createClass(PubSubProvider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return { pubSubCore: this.pubSubCore };
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var pubSubCore = this.pubSubCore;
      var nextPubSubCore = nextProps.pubSubCore;

      if (pubSubCore !== nextPubSubCore) {
        warnOfCoreChanges();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var children = this.props.children;

      return _react.Children.only(children);
    }
  }]);

  return PubSubProvider;
}(_react.Component);

exports.default = PubSubProvider;

PubSubProvider.propTypes = {
  pubSubCore: _pubSubShape2.default.isRequired,
  children: _react.PropTypes.element.isRequired
};

PubSubProvider.childContextTypes = {
  pubSubCore: _pubSubShape2.default.isRequired
};
